File: src/main/java/org/c3lang/intellij/completion/ConstCompletionContributor.java
package org.c3lang.intellij.completion;

import com.intellij.codeInsight.completion.CompletionParameters;
import com.intellij.codeInsight.completion.CompletionProvider;
import com.intellij.codeInsight.completion.CompletionResultSet;
import com.intellij.codeInsight.completion.InsertHandler;
import com.intellij.codeInsight.completion.InsertionContext;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.patterns.ElementPattern;
import com.intellij.psi.PsiElement;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.stubs.StubIndex;
import com.intellij.util.ProcessingContext;
import org.c3lang.intellij.C3Icons;
import org.c3lang.intellij.index.NameIndex;
import org.c3lang.intellij.intention.AddImportQuickFix;
import org.c3lang.intellij.psi.C3ConstDeclarationStmt;
import org.c3lang.intellij.psi.C3ModuleDefinition;
import org.c3lang.intellij.psi.C3PathConst;
import org.c3lang.intellij.psi.C3PathConstExpr;
import org.c3lang.intellij.psi.C3PsiElement;
import org.c3lang.intellij.psi.FullyQualifiedName;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;

import static com.intellij.patterns.PlatformPatterns.psiElement;
import static com.intellij.patterns.StandardPatterns.or;

@SuppressWarnings("DuplicatedCode")
public final class ConstCompletionContributor extends CompletionProvider<CompletionParameters>
{
    public static final ConstCompletionContributor INSTANCE = new ConstCompletionContributor();

    @SuppressWarnings("unused")
    private static final Logger LOG = Logger.getInstance(ConstCompletionContributor.class);

    private static final ElementPattern<PsiElement> PATTERN = or(
        psiElement().inside(C3PathConst.class)
    );

    private ConstCompletionContributor()
    {
    }

    @Override
    protected void addCompletions(
            @NotNull CompletionParameters parameters,
            @NotNull ProcessingContext context,
            @NotNull CompletionResultSet result)
    {
        if (!PATTERN.accepts(parameters.getPosition()) && !PATTERN.accepts(parameters.getOriginalPosition()))
        {
            return;
        }

        C3ModuleDefinition moduleDefinition = CompletionExtensionsKt.getModuleDefinition(parameters);
        if (moduleDefinition == null) return;

        C3PathConstExpr lookupTarget = CompletionExtensionsKt.siblingOf(parameters, C3PathConstExpr.class);
        if (lookupTarget == null) return;

        String lookupString = CompletionExtensionsKt.getLookupString(parameters, lookupTarget);
        var matcher = CompletionExtensionsKt.getMatcher(lookupString);

        var moduleName = moduleDefinition.getModuleName();
        TextRange elementRange = lookupTarget.getTextRange();
        Project project = parameters.getPosition().getProject();
        InsertHandler<LookupElement> insertHandler = new ConstInsertHandler(moduleDefinition, elementRange);

        for (String key : StubIndex.getInstance().getAllKeys(NameIndex.KEY, project))
        {
            if (!matcher.matches(key) && !key.isBlank()) continue;

            for (C3PsiElement element : StubIndex.getElements(
                    NameIndex.KEY,
                    key,
                    project,
                    GlobalSearchScope.allScope(project),
                    C3PsiElement.class))
            {
                if (!(element instanceof C3ConstDeclarationStmt constDeclaration)) continue;

                FullyQualifiedName fullyQualifiedName = constDeclaration.getFqName();
                LookupElementBuilder lookupElementBuilder = LookupElementBuilder
                    .create(constDeclaration, fullyQualifiedName.getFullName())
                    .withLookupStrings(List.of(
                        fullyQualifiedName.getFullName(),
                        fullyQualifiedName.getSuffixName(),
                        fullyQualifiedName.getName()
                    ))
                    .withLookupStrings(List.of(fullyQualifiedName.getFullName(), fullyQualifiedName.getName()))
                    .withIcon(C3Icons.Nodes.CONSTANT)
                    .withPresentableText(Objects.equals(fullyQualifiedName.getModule(), moduleName)
                        ? fullyQualifiedName.getName()
                        : fullyQualifiedName.getFullName())
                    .withInsertHandler(insertHandler);

                result.addElement(lookupElementBuilder);
            }
        }
    }

    @SuppressWarnings("DuplicatedCode")
    private static final class ConstInsertHandler implements InsertHandler<LookupElement>
    {
        private final C3ModuleDefinition moduleDefinition;
        private final TextRange range;

        private ConstInsertHandler(@NotNull C3ModuleDefinition moduleDefinition, @NotNull TextRange range)
        {
            this.moduleDefinition = moduleDefinition;
            this.range = range;
        }

        @Override
        public void handleInsert(@NotNull InsertionContext context, @NotNull LookupElement item)
        {
            PsiElement psiElement = item.getPsiElement();
            if (!(psiElement instanceof C3ConstDeclarationStmt element)) return;

            WriteCommandAction.runWriteCommandAction(context.getProject(), () -> {
                AddImportQuickFix.ImportAction imported =
                    AddImportQuickFix.Companion.addImportAsText(element, moduleDefinition);

                var importedModuleName = imported != null ? imported.getModuleName() : null;
                String textToInsert = moduleDefinition.textToInsert(importedModuleName, element);
                int endOffset = context.getEditor().getCaretModel().getOffset();

                context.getDocument().replaceString(
                    range.getStartOffset(),
                    endOffset,
                    textToInsert
                );
                context.getEditor().getCaretModel().moveToOffset(range.getStartOffset() + textToInsert.length());

                if (imported != null)
                {
                    imported.write(context.getDocument());
                }
            });
        }
    }
}

File: src/main/java/org/c3lang/intellij/completion/ConstCompletionContributor.kt
package org.c3lang.intellij.completion

import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.TextRange
import com.intellij.patterns.PlatformPatterns.or
import com.intellij.patterns.PlatformPatterns.psiElement
import com.intellij.psi.PsiErrorElement
import com.intellij.psi.codeStyle.NameUtil
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StubIndex
import org.c3lang.intellij.C3Icons
import org.c3lang.intellij.completion.PsiDocumentUtils.commitChanges
import org.c3lang.intellij.index.NameIndex
import org.c3lang.intellij.intention.AddImportQuickFix
import org.c3lang.intellij.psi.*

@Suppress("DuplicatedCode")
class ConstCompletionContributor : CompletionContributor() {
    private val pattern = or(
        psiElement(C3Types.CONST_IDENT)
    )

    override fun fillCompletionVariants(parameters: CompletionParameters, result: CompletionResultSet) {
        val originalPosition = parameters.originalPosition

        if (!pattern.accepts(originalPosition)) {
            return;
        }

        val parent = originalPosition?.parent
        if (parent is PsiErrorElement) {
            return
        }

        val completion = CompletionContext(parameters)

        val nameMatcher = NameUtil.buildMatcher(
            "*${completion.lookupString}*",
            NameUtil.MatchingCaseSensitivity.NONE
        )
        val insertHandler = ConstInsertHandler(
            moduleSection = completion.importProvider,
            range = completion.range
        )

        StubIndex.getInstance().getAllKeys(
            NameIndex.KEY,
            completion.project
        ).filter { nameMatcher.matches(it) || it.isBlank() }.flatMap { key ->
            StubIndex.getElements(
                NameIndex.KEY,
                key,
                completion.project,
                GlobalSearchScope.allScope(completion.project),
                C3PsiElement::class.java,
            )
        }.filterIsInstance<C3ConstDeclarationStmt>().forEach { const ->
            val fullyQualifiedName = FullyQualifiedName.from(const)

            val lookupElementBuilder = LookupElementBuilder
                .create(const, fullyQualifiedName.fullName)
                .withLookupStrings(listOf(fullyQualifiedName.fullName, fullyQualifiedName.name))
                .withPsiElement(const)
                .withIcon(C3Icons.Nodes.CONSTANT)
                .withPresentableText(fullyQualifiedName.fullName)
                .withInsertHandler(insertHandler)

            result.addElement(lookupElementBuilder)
        }
    }

    @Suppress("DuplicatedCode")
    private class ConstInsertHandler(
        private val moduleSection: C3ImportProvider,
        private val range: TextRange,
    ) : InsertHandler<LookupElement> {

        override fun handleInsert(context: InsertionContext, item: LookupElement) {
            val editor = context.editor
            val document = editor.document
            val element = item.psiElement as C3TypeName

            val moduleName = element.moduleName
            val moduleImported = moduleSection.imports.contains(element.moduleName)

            val textToInsert = if (moduleImported) {
                element.typeName.suffixName
            } else {
                element.typeName.fullName
            }

            val endOffset = editor.caretModel.offset

            document.replaceString(
                range.startOffset,
                endOffset,
                textToInsert
            )
            document.commitChanges(context)
            editor.caretModel.moveToOffset(range.startOffset + textToInsert.length)

            AddImportQuickFix.addImport(
                importIntention = moduleName,
                moduleSection = moduleSection,
                project = context.project
            )
        }

    }

    companion object {
        private val log = Logger.getInstance(
            ConstCompletionContributor::class.java.name
        )
    }
}
